### PR TITLE
#1088 transaction batches with no item lines

### LINKED ISSUE #1088: TransactionBatch with no ItemBatch
## Error in mSupply Mobile

**TypeError** in **MainActivity**
null is not an object (evaluating 'this.itemBatch.totalQuantity')

[View on Bugsnag](https://app.bugsnag.com/sustainable-solutions-nz-ltd/msupply-mobile/errors/5d4495fa0ebfc3001ad88139?event_id=5d4495fa004c28e4f2a20000&i=gh&m=ci)

## Stacktrace

    src/database/DataTypes/TransactionBatch.js:44 - value
    src/database/DataTypes/TransactionItem.js:247 - value
    src/database/DataTypes/TransactionItem.js:200 - value
    src/database/DataTypes/TransactionItem.js:178 - value
    src/pages/CustomerInvoicePage.js:106 - 
    src/database/UIDatabase.js:157 - readFileGeneric
    src/pages/CustomerInvoicePage.js:105 - onEndEditing

[View full stacktrace](https://app.bugsnag.com/sustainable-solutions-nz-ltd/msupply-mobile/errors/5d4495fa0ebfc3001ad88139?event_id=5d4495fa004c28e4f2a20000&i=gh&m=ci)

*Created automatically via Bugsnag*

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #1088 

## Change summary

- Migration code for adding a 0 quantity item batch to any transaction batch with no item batch

## Testing

- [ ] Upgrade an installed version, migration code runs, migration code doesn't crash
    - Can quite easily create a new apk for this from this branch. Make sure `package.json` is `2.3.6`
- [ ] Upgrade installed version with `TransactionBatch` with no `ItemBatch` relation, adds an `ItemBatch` related to the `TransactionBatch`'s with `0` quantity
    - Can quite easily do this by `this.database.delete('ItemBatch', this.database.objects('ItemBatch'))` in `componentDidMount` to delete all `ItemBatch` records, and any `TransactionBatch` records will not be related to any `ItemBatch`

### Related areas to think about
- What happens after this 'fix'
- Migration code files being moved effecting other areas - i.e. still runs

#### NOTE: 

Realm files exported before and after migration:
[AfterAndBeforeMigration.zip](https://github.com/openmsupply/mobile/files/3495195/AfterAndBeforeMigration.zip)

Branch: https://github.com/openmsupply/mobile/compare/235-with-transaction-delete-item-batch-code-no-fixer -- this branch has simple code to create the 'no `ItemBatch` for `TransactionBatch` situation

APKs: I have an apk built from the above branch, and another from this PR branch (too big to upload, message me :))

### CODE REVIEW

File: src/database/migration/v2.js [hunk 1]
```diff
@@ -188,6 +132,61 @@ const dataMigrations = [
...
+        // TransactionBatch are not directly related, just hold a reference. Find
+        // the actual Item record.
+        const items = database.objects('Item').filtered('id == $0', itemId);
+        // Double checking the itemId in an ItemBatch is actually related to an item.
+        // Early exit if not, something else is wrong
+        if (!items.length > 0) return;
+        const item = items[0];
+        // Create the new ItemBatch using as much detail from the TransactionBatch
+        // as possible
+        database.write(() => {
```
Comment (andreievg) [inline src/database/migration/v2.js]: hmm, usually we've been using createRecord.js for creating new record, it also links item.itemBatches to item batch (which technical is not needed here), and it's probably best to get costPrice and sellPrice from transaction batch, as you've done, I am happy with creating record the way you've done here.

File: src/database/migration/v2.js [hunk 2]
```diff
@@ -188,6 +132,61 @@ const dataMigrations = [
...
+            batch,
+            expiryDate,
+            packSize,
+            costPrice,
+            sellPrice,
+            donor,
+            item,
+          });
+          // Add the transaction batch into the new item batch
+          newItemBatch.transactionBatches.push(transactionBatch);
```
Comment (andreievg) [inline src/database/migration/v2.js]: can use addTransactionBatch, but not required

File: src/database/migration/v2.js [hunk 3]
```diff
@@ -188,6 +132,61 @@ const dataMigrations = [
...
+  {
+    version: '2.3.6',
+    migrate: database => {
+      // Find all the transactions with no item batches, if any
+      const transactionBatches = database
+        .objects('TransactionBatch')
+        .filtered('itemBatch == $0', null)
+        .snapshot();
+      // Do nothing if there aren't any.
+      if (!(transactionBatches.length > 0)) return;
```
Comment (andreievg) [inline src/database/migration/v2.js]: not a biggy, but this line is not required

File: src/database/migration/v2.js [hunk 4]
```diff
@@ -188,6 +132,61 @@ const dataMigrations = [
...
+      // For each of these item batchless-transaction batches, create a new
+      // 0 quantity item batch
+      transactionBatches.forEach(transactionBatch => {
+        if (!transactionBatch) return;
```
Comment (andreievg) [inline src/database/migration/v2.js]: this is probably excessive

File: src/database/migration/v2.js [hunk 5]
```diff
@@ -188,6 +132,61 @@ const dataMigrations = [
...
+            sellPrice,
+            donor,
+            item,
+          });
+          // Add the transaction batch into the new item batch
+          newItemBatch.transactionBatches.push(transactionBatch);
+          database.save('ItemBatch', newItemBatch);
+          // Update the TransactionBatch with the newly created ItemBatch
+          database.update('TransactionBatch', {
+            ...transactionBatch,
```
Comment (andreievg) [inline src/database/migration/v2.js]: for some reason i've never been a fan of full destructure in update, would rather just id. (if anything to be changed in this PR it's this)

File: src/database/migration/v2.js (hunk 2 above)
Comment (joshxg) [reply src/database/migration/v2.js]: fd6d00ff4f80fe222a2ac54eeaadb45369bc5055 used the unique one

File: src/database/migration/v2.js (hunk 3 above)
Comment (joshxg) [reply src/database/migration/v2.js]: d4989fd806f9a91a61716c13b8d3352c76b0927d

File: src/database/migration/v2.js (hunk 4 above)
Comment (joshxg) [reply src/database/migration/v2.js]: yeah .. especially with snapshotting, but realm lazy-but-live results are sometimes unpredictable heh

**Comment by mark-prins** [comment]
@joshxg @andreievg is it ok to merge this? We may need it in a release soon

**Comment by andreievg** [comment]
@mark-prins i am adding another bit of migration (today): https://github.com/openmsupply/mobile/issues/1102, will make 2.3.5 and 2.3.5-IC1 <- that one just for IC

**Comment by andreievg** [comment]
@joshxg, i'll rebase this to a special V2.3.6-IC1 branch

**Pull request merged**